shared/packages/peripherals/src/peripherals/midi.ts

```diff
@@ -0,0 +1,336 @@
+				return
+			}
+
+			// Figure out which MIDI devices have been unplugged since the last check.
+			const disconnectedDeviceIds = new Set(
+				Array.from(lastSeenDevices.keys()).filter((id) => {
+					return !devices.has(id)
+				})
+			)
+
+			// Figure out which MIDI devices are being seen now that weren't seen in the last completed poll.
+			for (const [id, { port, name }] of devices) {
+				const alreadySeen = lastSeenDevices.has(id)
+
+				if (alreadySeen) {
+					continue
+				}
+				// Tell the watcher about the discovered MIDI panel.
+				onKnownPeripheral(id, {
+					name: name,
+					type: PeripheralType.MIDI,
+					devicePath: `${port}`,
+				})
+			}
+
+			// Tell the watcher about disconnected MIDI devices.
+			for (const id of disconnectedDeviceIds) {
+				onKnownPeripheral(id, null)
+			}
+
+			// Update for the next iteration.
+			lastSeenDevices = devices
+		}
+
+		// Check for new or removed MIDI devices every 5 seconds.
+		const interval = setInterval(checkDevices, 5000)
+
+		// Do an initial check in 100 milliseconds.
+		setTimeout(checkDevices, 100)
+
+		return {
+			stop: () => {
+				clearInterval(interval)
+				PeripheralMIDI.Watching = false
+			},
+		}
+	}
+	private portIndex: number
+
+	public initializing = false
+	public connected = false
+
+	private Input?: MIDI.Input
+	private Output?: MIDI.Output
+
+	private _info: PeripheralInfo | undefined
+	private sentKeyDisplay: { [identifier: string]: KeyDisplay } = {}
+
+	private seenKeys = new Map<
+		string,
+		{
+			channel: number
+			keyNumber: number
+			state: boolean
+		}
+	>()
+	private intervals = new Map<
+		string,
+		{
+			hash: string
+		}
+	>()
+
+	constructor(log: LoggerLike, id: string, portIndexString: string) {
+		super(log, id)
+		this.portIndex = parseInt(portIndexString, 10)
+	}
+
+	async init(): Promise<void> {
+		try {
+			this.initializing = true
+
+			const Input = new MIDI.Input()
+
+			this._info = {
+				name: Input.getPortName(this.portIndex),
+				gui: {
+					type: PeripheralType.MIDI,
+				},
+			}
+			// MIDI devices connected:
+			Input.on('message', (_, message: MIDI.MidiMessage) => {
+				this.onMidiMessage(message)
+			})
+			// @ts-expect-error error event not in typings
+			Input.on('error', (err) => this.log.error(err))
+
+			Input.openPort(this.portIndex)
+			this.Input = Input
+
+			// Find corresponding output:
+			const Output = new MIDI.Output()
+			const outputPortCount = Output.getPortCount()
+			for (let i = 0; i < outputPortCount; i++) {
+				if (Output.getPortName(i) === this.info.name) {
+					Output.openPort(i)
+					this.Output = Output
+				}
+			}
+
+			this.connected = true
+
+			this.initializing = false
+		} catch (e) {
+			this.initializing = false
+			throw e
+		}
+	}
+	get info(): PeripheralInfo {
+		if (!this._info) throw new Error('Peripheral not initialized')
+		return this._info
+	}
+	async _setKeyDisplay(identifier: string, keyDisplay: KeyDisplay, force = false): Promise<void> {
```

Comment:
Does this have a purpose with MIDI devices, or can it be removed?

Reply:
Yes, some midi devices have lights on buttons, this is used to to indicate status for those buttons.